src/OpenCL.cpp

```diff
@@ -403,36 +344,48 @@ void OpenCL_Network::forward(const std::vector<net_t>& input,
                              std::vector<net_t>& output) {
     constexpr auto width = 19;
     constexpr auto height = 19;
+    constexpr auto tiles = (width + 1)*(height + 1) / 4;
     constexpr auto one_plane = width * height * sizeof(net_t);
 
     opencl.ensure_thread_initialized();
 
     if (!opencl_thread_data.m_buffers_allocated) {
-        auto maxInBufferSize = 0;
-        auto maxMergeSize = 0;
+        unsigned int max_channels = 0;
         for (const auto& layer : m_layers) {
-            auto channelGroups = layer.channels / (layer.channels % 8 ? 2 : 8);
-            maxMergeSize = std::max<int>(maxMergeSize,
-                                         layer.outputs * channelGroups);
-            maxInBufferSize = std::max<int>(maxInBufferSize, layer.channels);
+            max_channels = std::max(max_channels,
+                    std::max(layer.channels, layer.outputs));
         }
-        const auto alloc_inSize = one_plane *  maxInBufferSize;
-        const auto alloc_mergeSize = one_plane * maxMergeSize;
+
+        const auto mwg = opencl.m_sgemm_tuners.mwg;
+        const auto nwg = opencl.m_sgemm_tuners.nwg;
+        const auto vwm = opencl.m_sgemm_tuners.vwm;
+        const auto vwn = opencl.m_sgemm_tuners.vwn;
+
+        const auto m_ceil = lcm(lcm(max_channels, mwg), vwm);
+        const auto n_ceil = lcm(lcm(tiles, nwg), vwn);
+
+        const auto alloc_inSize = m_ceil * m_ceil *  max_channels * sizeof(net_t);
+        const auto alloc_vm_size = WINOGRAD_TILE * m_ceil * n_ceil * sizeof(net_t);
+
+        auto v_zeros = std::vector<float>(alloc_vm_size);
 
         opencl_thread_data.m_inBuffer = cl::Buffer(
             CL_MEM_READ_WRITE, alloc_inSize);
         opencl_thread_data.m_tmpBuffer = cl::Buffer(
             CL_MEM_READ_WRITE, alloc_inSize);
         opencl_thread_data.m_residualBuffer = cl::Buffer(
             CL_MEM_READ_WRITE, alloc_inSize);
-        opencl_thread_data.m_mergeBuffer = cl::Buffer(
-            CL_MEM_READ_WRITE | CL_MEM_HOST_NO_ACCESS, alloc_mergeSize);
+        opencl_thread_data.m_VBuffer = cl::Buffer(
+            CL_MEM_READ_WRITE | CL_MEM_HOST_NO_ACCESS | CL_MEM_COPY_HOST_PTR, alloc_vm_size, v_zeros.data(), nullptr);
```

Comment:
Does this actually need to be zeroed? It feels weird because it's not going to be used and filled up. Has to do with padding?

Reply:
This is for zero padding. It's zeroed at the beginning once and the padding is not touched during the operations.